### DISCUSSION TITLE
About polygon mesh showes texture

### DISCUSSION THREAD

**Discussion by ZiguoAtGitHub** [marcomusy/vedo#787] [Q&A] [answered]
Hi Marco,
I want to map a jpg onto polygon meshes with texture function, but I could not get what I want correctly. 
I try modify the parameters and add light, but all of them doesn't work. 
Here is my example code. 
```
"""Manually build a mesh from points and faces"""
from vedo import Mesh, show, Light

verts = [(50,50,50), (70,40,50), (50,40,80), (80,70,50)]
faces = [(0,1,2), (2,1,3), (1,0,3)]
# (the first triangle face is formed by vertex 0, 1 and 2)

# Build the polygonal Mesh object:
mesh = Mesh([verts, faces]).texture('Exposed-Aggregate-Architextures.jpg',edge_clamp=True,interpolate=False,repeat=False)
#mesh.backcolor('violet').linecolor('tomato').linewidth(2)
labs = mesh.labels('id').c('black')


# retrieve them as numpy arrays
print('points():', mesh.points())
print('faces() :', mesh.faces())
light=Light(pos=(80,40,70),focal_point=(50,50,50),c='white',angle=90)

show(mesh, labs, __doc__, viewup='z', axes=1).close()
#show(mesh, labs,light,  __doc__, viewup='z', axes=1).close()
```

**Comment by marcomusy** [comment]
Hi there are 2 issues in one here :)
About the texture you need to provide a set of texture coordinates that map vertices to position on the image to map (u,v) coordinates.
About lights it seems to work for me.. try experiment with pressing `o` and `O` (to disable)

**Comment by ZiguoAtGitHub** [comment]
Hi Marco,
Yes, the light works as it ought to do. “doesn't work" means that I supposed the texture had no effects  were caused by no light on it. I am sorry for my poor description and misleading you. 🤦‍
As your hints, I got what I want after I added the parameter tcoords=[(0,0),(0,1),(1,0),(0,0)] of the texture method.
Thanks! I think I get something about the texture how to work. :)
Here is texture effect.
![image](https://user-images.githubusercontent.com/116484589/211442645-ef238fb9-5228-4d4e-bf95-0c3db415d244.png)

**Comment by ZiguoAtGitHub** [comment]
Hi    Marco @marcomusy ,
Some cases are more complicated.
I have three polygon surfaces, here is one example and there are more than 1000 surfaces in my real case, which connected with each other. How can I get the same texture result for every surface.
I tried to modify the coordinates of tcoords and failed. Would you have some ideas? Any ideas are appreicated.
Here is my screenshot.
![image](https://user-images.githubusercontent.com/116484589/211536487-de9a5471-d2d0-462f-b688-92458f29a462.png)
Here is the code.

```
"""Manually build a mesh from points and faces"""
from vedo import Mesh, show, Light

verts = [(0,0,0), (10,0,0), (10,0,20), (0,0,20),(0,10,0),(0,10,20),(14,14,20)]
faces = [(0,1,2,3), (0,3,5,4),(2,3,5,6)]
# (the first triangle face is formed by vertex 0, 1 and 2)

# Build the polygonal Mesh object:
mesh = Mesh([verts, faces]).texture('concrete-texture-1445436.jpg',tcoords=[(0,0),(0,1),(1,1),(1,0),(1,0),(1,1),(0,0)],edge_clamp=True,interpolate=False,repeat=True)#
#mesh.backcolor('violet').linecolor('tomato').linewidth(2)
labs = mesh.labels('id').c('black')
# retrieve them as numpy arrays
print('points():', mesh.points())
print('faces() :', mesh.faces())

show(mesh, labs, __doc__, viewup='z', axes=1).close()
#show(mesh, labs,light,  __doc__, viewup='z', axes=1).close()

```
Thanks again.

**Reply by marcomusy** [reply]
Hi Ziguo there is not "cheap and automatic" way of doing such a thing.. you need to define the texture coordinates of the uv map ,by "unwrapping the mesh onto a plane:

![](https://cdn.conceptartempire.com/images/02/6627/00-featured-uv-unwrapping-blender-sample.jpg)

see eg:
https://blenderartists.org/t/how-to-create-a-texture-from-a-mesh/1118259

what you can do automatically is to attach a texture to a surface by means of fitting the best plane to  it.
in case of a simple shape like a cube you will need to compute normals (`mycube.compute_normals(cells=True, points=True)`) before calling `texture()`.

**Reply by ZiguoAtGitHub** [reply]
> Hi Ziguo there is not "cheap and automatic" way of doing such a thing.. you need to define the texture coordinates of the uv map ,by "unwrapping the mesh onto a plane:
> 
> ![](https://camo.githubusercontent.com/c274b05eb1e1aa7e533c5e814007d3d17cb37b7769fdb97a1ddb7cddf077f41d/68747470733a2f2f63646e2e636f6e63657074617274656d706972652e636f6d2f696d616765732f30322f363632372f30302d66656174757265642d75762d756e7772617070696e672d626c656e6465722d73616d706c652e6a7067)
> 
> see eg: https://blenderartists.org/t/how-to-create-a-texture-from-a-mesh/1118259
> 
> what you can do automatically is to attach a texture to a surface by means of fitting the best plane to it. in case of a simple shape like a cube you will need to compute normals (`mycube.compute_normals(cells=True, points=True)`) before calling `texture()`.

Hi Marco,
Thanks for your kindly reply!
Honestly, I am a beginner in computer 3D visualization area. After I browsed the website of you mentioned I have more deeper understanding about texture.

The lazy also the only way I can figure out by now is creating individual mesh and mapping texture for each face to get the expected effect, showed as following picture. I don't want to tell you how long to plot. 😂
![image](https://user-images.githubusercontent.com/116484589/211734593-95bc5f37-9c91-49a3-8a4b-e03caa23c1de.png)

The following figure is using the fast way, create one mesh for all faces. As you see, the effect in general view is not too bad but the details ....
![image](https://user-images.githubusercontent.com/116484589/211738180-cbed7e65-a93d-488e-adb8-86779cb5da53.png)

**I want to try to get what I want in next days although there is no any clues in my head. Do you think "unwrapping the mesh onto a plane" is possible toward my problem? Any ideas and comments will be deeply appreciated!**

**Comment by ZiguoAtGitHub** [comment]
I just failed twice:
1. Creating single mesh for each face, mapping the texture and then merge them into one mesh. However the texture lost after merged.
2. Creating single mesh for each face, mapping the texture and then creating one Assembly with all meshed. However the plotting speed is much tearful. 😂

**Comment by ZiguoAtGitHub** [comment, marked as answer]
I got one doable way.
To duplicate the share points between connecting faces so as to no vertices shared by different faces. And then the tcoords of texture could be genertaed with [(0,0),(1,0),(1,1),(0,1)] for four-vertices face or  [(0,0),(1,0),(1,1)] for three-vertices face.
Here is the result. It looks good enough to me.
![image](https://user-images.githubusercontent.com/116484589/212464347-1e06801c-d443-4fe1-8011-84238f12bf47.png)
